File: tests/load/conversation_explain.py
#!/usr/bin/env python3
"""Read-only PostgreSQL EXPLAIN gate for imported-scale conversation paging."""

from __future__ import annotations

import argparse
import json
import os
import pathlib
import re
import shutil
import subprocess
import sys
from typing import Any


SAFE_UUID = re.compile(r"^[0-9A-Fa-f-]{36}$")


class GateFailure(RuntimeError):
    pass


def psql(database_url: str, sql: str, timeout_ms: int) -> str:
    environment = os.environ.copy()
    environment["PGDATABASE"] = database_url
    environment["PGOPTIONS"] = (
        f"{environment.get('PGOPTIONS', '')} -c default_transaction_read_only=on "
        f"-c statement_timeout={timeout_ms}"
    ).strip()
    result = subprocess.run(
        ["psql", "-X", "-q", "-A", "-t", "--no-psqlrc"],
        input=sql,
        capture_output=True,
        text=True,
        env=environment,
        timeout=max(30, timeout_ms // 1000 + 10),
        check=False,
    )
    if result.returncode:
        raise GateFailure((result.stderr.strip().splitlines() or ["psql failed"])[-1])
    return result.stdout.strip()


def scalar(database_url: str, sql: str, timeout_ms: int) -> str:
    rows = psql(database_url, sql, timeout_ms).splitlines()
    return rows[-1].strip() if rows else ""


def nodes(plan: dict[str, Any]) -> list[dict[str, Any]]:
    return [plan] + [item for child in plan.get("Plans", []) for item in nodes(child)]


def explain(database_url: str, name: str, query: str, timeout_ms: int) -> dict[str, Any]:
    document = json.loads(
        psql(
            database_url,
            f"EXPLAIN (ANALYZE, BUFFERS, FORMAT JSON, TIMING OFF) {query};",
            timeout_ms,
        )
    )[0]
    plan_nodes = nodes(document["Plan"])
    forbidden_scans = sorted(
        {
            str(node.get("Relation Name"))
            for node in plan_nodes
            if node.get("Node Type") == "Seq Scan"
            and str(node.get("Relation Name", "")).startswith(
                (
                    "request_records",
                    "conversation_key_clusters",
                    "conversation_observations",
                    "conversation_edges",
                )
            )
        }
    )
    return {
        "name": name,
        "execution_time_ms": document.get("Execution Time", 0.0),
        "returned_rows": document["Plan"].get("Actual Rows", 0),
        "indexes": sorted(
            {
                str(node["Index Name"])
                for node in plan_nodes
                if node.get("Index Name") is not None
            }
        ),
        "forbidden_sequential_scans": forbidden_scans,
        "plan": document["Plan"],
    }


def arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--database-url-file", type=pathlib.Path)
    parser.add_argument("--output", type=pathlib.Path)
    parser.add_argument("--min-request-rows", type=int, default=110_000)
    parser.add_argument("--max-execution-ms", type=float, default=250.0)
    parser.add_argument("--statement-timeout-ms", type=int, default=30_000)
    return parser.parse_args()


def main() -> int:
    args = arguments()
    if shutil.which("psql") is None:
        print(json.dumps({"passed": False, "error": "psql is not installed"}), file=sys.stderr)
        return 3
    try:
        database_url = (
            args.database_url_file.read_text().strip()
            if args.database_url_file
            else os.environ.get("MTC_BENCH_DATABASE_URL", "").strip()
        )
        if not database_url:
            raise GateFailure("MTC_BENCH_DATABASE_URL or --database-url-file is required")
        request_rows = int(
            scalar(database_url, "SELECT COUNT(*) FROM request_records;", args.statement_timeout_ms)
        )
        projection_rows = int(
            scalar(
                database_url,
                "SELECT COUNT(*) FROM conversation_key_clusters;",
                args.statement_timeout_ms,
            )
        )
        sample = scalar(
            database_url,
            "SELECT key_id || '|' || cluster_id FROM conversation_key_clusters "
            "ORDER BY request_count DESC, updated_at DESC LIMIT 1;",
            args.statement_timeout_ms,
        ).split("|")
        if len(sample) != 2 or not all(SAFE_UUID.fullmatch(value) for value in sample):
            raise GateFailure("no valid conversation projection is available")
        key_id, cluster_id = sample
        cursor = scalar(
            database_url,
            "SELECT created_at::text || '|' || id FROM request_records "
            f"WHERE key_id = '{key_id}' AND conversation_cluster_id = '{cluster_id}' "
            "ORDER BY created_at DESC, id DESC OFFSET 99 LIMIT 1;",
            args.statement_timeout_ms,
        ).split("|")
        if len(cursor) != 2 or not cursor[0].isdigit() or not SAFE_UUID.fullmatch(cursor[1]):
            cursor = ["9223372036854775807", "ffffffff-ffff-ffff-ffff-ffffffffffff"]
        before_created_at, before_request_id = cursor
        queries = [
            (
                "conversation_list_first_page",
                "SELECT p.cluster_id, c.explicit_session_id, p.updated_at, p.request_count, "
                "p.candidate_edge_count FROM conversation_key_clusters p "
                "JOIN conversation_clusters c ON c.id = p.cluster_id "
                f"WHERE p.key_id = '{key_id}' "
                "ORDER BY p.updated_at DESC, p.cluster_id DESC LIMIT 50",
            ),
            (
                "conversation_detail_membership",
                "SELECT p.cluster_id, p.request_count FROM conversation_key_clusters p "
                f"WHERE p.key_id = '{key_id}' AND p.cluster_id = '{cluster_id}'",
            ),
            (
                "conversation_detail_first_page",
                "SELECT id, created_at, protocol, model, status_code FROM request_records "
                f"WHERE key_id = '{key_id}' AND conversation_cluster_id = '{cluster_id}' "
                "ORDER BY created_at DESC, id DESC LIMIT 201",
            ),
            (
                "conversation_detail_older_page",
                "SELECT id, created_at, protocol, model, status_code FROM request_records "
                f"WHERE key_id = '{key_id}' AND conversation_cluster_id = '{cluster_id}' "
                f"AND (created_at < {before_created_at} OR "
                f"(created_at = {before_created_at} AND id < '{before_request_id}')) "
                "ORDER BY created_at DESC, id DESC LIMIT 201",
            ),
        ]
        results = [explain(database_url, name, query, args.statement_timeout_ms) for name, query in queries]
        checks: list[dict[str, Any]] = [
            {
                "name": "imported request scale",
                "actual": request_rows,
                "expected_minimum": args.min_request_rows,
                "passed": request_rows >= args.min_request_rows,
            },
            {
                "name": "conversation projections populated",
                "actual": projection_rows,
                "expected_minimum": 1,
                "passed": projection_rows > 0,
            },
        ]
        for result in results:
            checks.extend(
                [
                    {
                        "name": f"{result['name']} latency",
                        "actual": result["execution_time_ms"],
                        "expected_maximum": args.max_execution_ms,
                        "passed": result["execution_time_ms"] <= args.max_execution_ms,
                    },
                    {
                        "name": f"{result['name']} indexed",
                        "actual": result["forbidden_sequential_scans"],
                        "expected": [],
                        "passed": not result["forbidden_sequential_scans"],
                    },
                ]
            )
        report = {
            "schema_version": 1,
            "dataset": {"request_rows": request_rows, "projection_rows": projection_rows},
            "sample": {"key_id": key_id, "cluster_id": cluster_id},
            "thresholds": {
                "min_request_rows": args.min_request_rows,
                "max_execution_ms": args.max_execution_ms,
            },
            "results": results,
            "checks": checks,
            "passed": all(check["passed"] for check in checks),
        }
        if args.output:
            args.output.parent.mkdir(parents=True, exist_ok=True)
            args.output.write_text(json.dumps(report, indent=2, sort_keys=True) + "\n")
        print(json.dumps({"passed": report["passed"], "checks": checks}, indent=2))
        return 0 if report["passed"] else 2
    except (OSError, ValueError, json.JSONDecodeError, subprocess.SubprocessError, GateFailure) as error:
        print(json.dumps({"passed": False, "error": str(error)}), file=sys.stderr)
        return 3


if __name__ == "__main__":
    raise SystemExit(main())

File: tests/load/conversation_explain.ts
#!/usr/bin/env node
/** Read-only PostgreSQL EXPLAIN gate for imported-scale conversation paging. */

import { spawnSync } from "node:child_process";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname } from "node:path";

const SAFE_UUID = /^[0-9A-Fa-f-]{36}$/;

export class GateFailure extends Error {}

type JsonObject = Record<string, any>;

interface Arguments {
  databaseUrlFile?: string;
  output?: string;
  minRequestRows: number;
  maxExecutionMs: number;
  statementTimeoutMs: number;
}

function parseNumber(flag: string, value: string | undefined): number {
  if (value === undefined || !Number.isFinite(Number(value))) {
    throw new GateFailure(`${flag} requires a number`);
  }
  return Number(value);
}

function argumentsFrom(argv = process.argv.slice(2)): Arguments {
  const args: Arguments = {
    minRequestRows: 110_000,
    maxExecutionMs: 250,
    statementTimeoutMs: 30_000,
  };
  for (let index = 0; index < argv.length; index += 1) {
    const flag = argv[index];
    const value = argv[index + 1];
    if (flag === "--database-url-file") args.databaseUrlFile = value;
    else if (flag === "--output") args.output = value;
    else if (flag === "--min-request-rows") args.minRequestRows = parseNumber(flag, value);
    else if (flag === "--max-execution-ms") args.maxExecutionMs = parseNumber(flag, value);
    else if (flag === "--statement-timeout-ms") args.statementTimeoutMs = parseNumber(flag, value);
    else if (flag === "--help" || flag === "-h") {
      console.log("Usage: conversation_explain.ts [--database-url-file PATH] [--output PATH] [--min-request-rows N] [--max-execution-ms N] [--statement-timeout-ms N]");
      process.exit(0);
    } else throw new GateFailure(`unrecognized argument: ${flag}`);
    index += 1;
  }
  return args;
}

function executableExists(name: string): boolean {
  return (process.env.PATH ?? "").split(":").some((directory) => existsSync(`${directory}/${name}`));
}

export function psql(databaseUrl: string, sql: string, timeoutMs: number): string {
  const result = spawnSync("psql", ["-X", "-q", "-A", "-t", "--no-psqlrc"], {
    input: sql,
    encoding: "utf8",
    timeout: Math.max(30_000, timeoutMs + 10_000),
    env: {
      ...process.env,
      PGDATABASE: databaseUrl,
      PGOPTIONS: `${process.env.PGOPTIONS ?? ""} -c default_transaction_read_only=on -c statement_timeout=${timeoutMs}`.trim(),
    },
  });
  if (result.error) throw new GateFailure(result.error.message);
  if (result.status !== 0) {
    const lines = (result.stderr ?? "").trim().split(/\r?\n/u).filter(Boolean);
    throw new GateFailure(lines.at(-1) ?? "psql failed");
  }
  return (result.stdout ?? "").trim();
}

function scalar(databaseUrl: string, sql: string, timeoutMs: number): string {
  return psql(databaseUrl, sql, timeoutMs).split(/\r?\n/u).at(-1)?.trim() ?? "";
}

function nodes(plan: JsonObject): JsonObject[] {
  return [plan, ...(Array.isArray(plan.Plans) ? plan.Plans.flatMap((child: JsonObject) => nodes(child)) : [])];
}

function explain(databaseUrl: string, name: string, query: string, timeoutMs: number): JsonObject {
  const parsed = JSON.parse(psql(databaseUrl, `EXPLAIN (ANALYZE, BUFFERS, FORMAT JSON, TIMING OFF) ${query};`, timeoutMs)) as JsonObject[];
  const document = parsed[0]!;
  const plan = document.Plan as JsonObject;
  const planNodes = nodes(plan);
  const relations = ["request_records", "conversation_key_clusters", "conversation_observations", "conversation_edges"];
  return {
    name,
    execution_time_ms: document["Execution Time"] ?? 0,
    returned_rows: plan["Actual Rows"] ?? 0,
    indexes: [...new Set(planNodes.map((node) => node["Index Name"]).filter((value): value is string => value !== undefined))].sort(),
    forbidden_sequential_scans: [...new Set(planNodes.filter((node) => node["Node Type"] === "Seq Scan" && relations.some((relation) => String(node["Relation Name"] ?? "").startsWith(relation))).map((node) => String(node["Relation Name"])))].sort(),
    plan,
  };
}

export function main(argv = process.argv.slice(2)): number {
  let args: Arguments;
  try {
    args = argumentsFrom(argv);
  } catch (error) {
    console.error(JSON.stringify({ passed: false, error: error instanceof Error ? error.message : String(error) }));
    return 3;
  }
  if (!executableExists("psql")) {
    console.error(JSON.stringify({ passed: false, error: "psql is not installed" }));
    return 3;
  }
  try {
    const databaseUrl = (args.databaseUrlFile ? readFileSync(args.databaseUrlFile, "utf8") : process.env.MTC_BENCH_DATABASE_URL ?? "").trim();
    if (!databaseUrl) throw new GateFailure("MTC_BENCH_DATABASE_URL or --database-url-file is required");
    const requestRows = Number.parseInt(scalar(databaseUrl, "SELECT COUNT(*) FROM request_records;", args.statementTimeoutMs), 10);
    const projectionRows = Number.parseInt(scalar(databaseUrl, "SELECT COUNT(*) FROM conversation_key_clusters;", args.statementTimeoutMs), 10);
    const sample = scalar(databaseUrl, "SELECT key_id || '|' || cluster_id FROM conversation_key_clusters ORDER BY request_count DESC, updated_at DESC LIMIT 1;", args.statementTimeoutMs).split("|");
    if (sample.length !== 2 || !sample.every((value) => SAFE_UUID.test(value))) throw new GateFailure("no valid conversation projection is available");
    const [keyId, clusterId] = sample as [string, string];
    let cursor = scalar(databaseUrl, `SELECT created_at::text || '|' || id FROM request_records WHERE key_id = '${keyId}' AND conversation_cluster_id = '${clusterId}' ORDER BY created_at DESC, id DESC OFFSET 99 LIMIT 1;`, args.statementTimeoutMs).split("|");
    if (cursor.length !== 2 || !/^\d+$/u.test(cursor[0] ?? "") || !SAFE_UUID.test(cursor[1] ?? "")) cursor = ["9223372036854775807", "ffffffff-ffff-ffff-ffff-ffffffffffff"];
    const [beforeCreatedAt, beforeRequestId] = cursor as [string, string];
    const queries: Array<[string, string]> = [
      ["conversation_list_first_page", `SELECT p.cluster_id, c.explicit_session_id, p.updated_at, p.request_count, p.candidate_edge_count FROM conversation_key_clusters p JOIN conversation_clusters c ON c.id = p.cluster_id WHERE p.key_id = '${keyId}' ORDER BY p.updated_at DESC, p.cluster_id DESC LIMIT 50`],
      ["conversation_detail_membership", `SELECT p.cluster_id, p.request_count FROM conversation_key_clusters p WHERE p.key_id = '${keyId}' AND p.cluster_id = '${clusterId}'`],
      ["conversation_detail_first_page", `SELECT id, created_at, protocol, model, status_code FROM request_records WHERE key_id = '${keyId}' AND conversation_cluster_id = '${clusterId}' ORDER BY created_at DESC, id DESC LIMIT 201`],
      ["conversation_detail_older_page", `SELECT id, created_at, protocol, model, status_code FROM request_records WHERE key_id = '${keyId}' AND conversation_cluster_id = '${clusterId}' AND (created_at < ${beforeCreatedAt} OR (created_at = ${beforeCreatedAt} AND id < '${beforeRequestId}')) ORDER BY created_at DESC, id DESC LIMIT 201`],
    ];
    const results = queries.map(([name, query]) => explain(databaseUrl, name, query, args.statementTimeoutMs));
    const checks: JsonObject[] = [
      { name: "imported request scale", actual: requestRows, expected_minimum: args.minRequestRows, passed: requestRows >= args.minRequestRows },
      { name: "conversation projections populated", actual: projectionRows, expected_minimum: 1, passed: projectionRows > 0 },
    ];
    for (const result of results) checks.push(
      { name: `${result.name} latency`, actual: result.execution_time_ms, expected_maximum: args.maxExecutionMs, passed: result.execution_time_ms <= args.maxExecutionMs },
      { name: `${result.name} indexed`, actual: result.forbidden_sequential_scans, expected: [], passed: result.forbidden_sequential_scans.length === 0 },
    );
    const report = { schema_version: 1, dataset: { request_rows: requestRows, projection_rows: projectionRows }, sample: { key_id: keyId, cluster_id: clusterId }, thresholds: { min_request_rows: args.minRequestRows, max_execution_ms: args.maxExecutionMs }, results, checks, passed: checks.every((check) => check.passed) };
    if (args.output) { mkdirSync(dirname(args.output), { recursive: true }); writeFileSync(args.output, `${JSON.stringify(report, null, 2)}\n`); }
    console.log(JSON.stringify({ passed: report.passed, checks }, null, 2));
    return report.passed ? 0 : 2;
  } catch (error) {
    console.error(JSON.stringify({ passed: false, error: error instanceof Error ? error.message : String(error) }));
    return 3;
  }
}

if (import.meta.url === `file://${process.argv[1]}`) process.exitCode = main();
